apisix/plugins/loggly.lua

```diff
@@ -0,0 +1,256 @@
+    local metadata = plugin.plugin_metadata(plugin_name)
+    local entry
+
+    if metadata and metadata.value.log_format
+       and core.table.nkeys(metadata.value.log_format) > 0
+    then
+        entry = log_util.get_custom_format_log(ctx, metadata.value.log_format)
+    else
+        entry = log_util.get_full_log(ngx, conf)
+    end
+
+    -- generate rfc5424 compliant syslog event
+    local json_str, err = core.json.encode(entry)
+    if not json_str then
+        core.log.error('error occurred while encoding the data: ', err)
+        return nil
+    end
+
+    local timestamp = log_util.get_rfc3339_zulu_timestamp()
+    local taglist = {}
+    if conf.tags then
+        for i = 1, #conf.tags do
+            core.table.insert(taglist, "tag=\"" .. conf.tags[i] .. "\"")
+        end
+    end
+    local message = {
+        -- facility LOG_USER - random user level message
+        "<".. tostring(8 + severity[conf.severity:upper()]) .. ">1",-- <PRIVAL>1
+        timestamp,                                                  -- timestamp
+        ctx.var.host or "-",                                        -- hostname
+        "apisix",                                                   -- appname
+        ctx.var.pid,                                                -- proc-id
+        "-",                                                        -- msgid
+        "[" .. conf.customer_token .. "@41058 " .. tab_concat(taglist, " ") .. "]",
+        json_str
+    }
+
+    return tab_concat(message, " ")
+end
+
+
+local function send_data_over_udp(message)
+    local metadata = plugin.plugin_metadata(plugin_name)
+    core.log.info("metadata: ", core.json.delay_encode(metadata))
+
+    if not metadata then
+        core.log.info("received nil metadata: using metadata defaults: ",
+                            core.json.delay_encode(defaults, true))
+        metadata = {}
+        metadata.value = defaults
+    end
+    local err_msg
+    local res = true
+    local sock = udp()
+    local host, port = metadata.value.host, metadata.value.port
+    sock:settimeout(metadata.value.timeout)
+
+    core.log.info("sending a batch logs to ", host, ":", port)
+
+    local ok, err = sock:setpeername(host, port)
+
+    if not ok then
+        return false, "failed to connect to UDP server: host[" .. host
+                    .. "] port[" .. tostring(port) .. "] err: " .. err
+    end
+
+    ok, err = sock:send(message)
+    if not ok then
+        res = false
+        err_msg = "failed to send data to UDP server: host[" .. host
+                  .. "] port[" .. tostring(port) .. "] err:" .. err
+    end
+
+    ok, err = sock:close()
+    if not ok then
+        core.log.error("failed to close the UDP connection, host[",
+                        host, "] port[", port, "] ", err)
+    end
+
+    return res, err_msg
+end
+
+
+local function handle_log(entries)
+    local ok, err
+    for i = 1, #entries do
+        ok, err = send_data_over_udp(entries[i])
```

Comment:
do we need to check the return value?

Reply:
Look like we need to remove the sent entries? Otherwise, they will be retried.

Reply:
Yes, that's an issue here. The way batchprocessor is written, it is meant for consuming each batch in one go (transactional). But in the case of loggly, only bulk events can be sent through HTTP endpoints. So I can think of three potential resolution

1. if err and i==1 -> then return false else true | if some entries has been sent that's ok. we are loosing the rest of the entries. [some entries will be dropped]
2. use the current behaviour. where some entries might be retried again [duplocation].
3. Update batch processor to have this fine-grained control
https://github.com/apache/apisix/blob/a813b5aa8980dd8fa5780af351b6b47b5219d0b9/apisix/utils/batch-processor.lua#L66-L89

update 
```lua
local ok, err = self.func(batch.entries, self.batch_max_size)

-- to this
local ok, err, n = self.func(batch.entries, self.batch_max_size)
 -- shorten self.entries based on n
```

which one do you think would be good? cc @spacewander

Reply:
here n denotes the number of processed entries, and the rest of the entries should be retried.

Reply:
cc @spacewander